# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю.

Выполнили студенты гр. 2383 Миненок Алиса и Мордасов Евгений. Вариант №27.

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».



## Основные теоретические положения


В данной работе исследуется система двух непрерывных случайных величин $(X, Y)$ на основе двумерной выборки объема $n$.

#### 1. Корреляционная зависимость и корреляционная таблица
**Корреляционная зависимость** — это статистическая зависимость, при которой изменение значения одной случайной величины приводит к изменению математического ожидания (среднего значения) другой.
Для непрерывных признаков строится **двумерный интервальный вариационный ряд**, который оформляется в виде корреляционной таблицы. В ячейках таблицы указываются частоты совместного появления признаков $n_{ij}$ (где $i$ — номер интервала по $X$, $j$ — номер интервала по $Y$). Маргинальные (одномерные) частоты обозначаются как $n_x$ и $n_y$, при этом $\sum n_x = \sum n_y = n$.

#### 2. Выборочный коэффициент корреляции ($r_{xy}$)
Служит безразмерной количественной мерой тесноты и направления линейной связи между случайными величинами. Значение коэффициента всегда лежит в пределах $[-1; 1]$. В работе он вычисляется двумя тождественными методами:
1. **По стандартной формуле для сгруппированных данных:**
   $$r_{xy} = \frac{\sum \sum n_{ij} x_i y_j - n \bar{x} \bar{y}}{n \sigma_x \sigma_y}$$
   где $x_i, y_j$ — середины интервалов, $\bar{x}, \bar{y}$ — выборочные средние, $\sigma_x, \sigma_y$ — выборочные средние квадратические отклонения.
2. **Методом условных вариант (для упрощения вычислений):**
   Вводятся новые переменные (условные варианты): $u_i = \frac{x_i - C_x}{h_x}$ и $v_j = \frac{y_j - C_y}{h_y}$, где $C_x$ и $C_y$ — ложные нули (середины модальных интервалов), $h_x, h_y$ — шаги интервалов.
   $$r_{xy} = \frac{\sum \sum n_{ij} u_i v_j - n \bar{u} \bar{v}}{n \sigma_u \sigma_v}$$

#### 3. Интервальная оценка коэффициента корреляции
Выборочный коэффициент $r_{xy}$ является точечной оценкой. Для построения доверительного интервала, который с заданной надежностью $\gamma$ накрывает истинный генеральный коэффициент корреляции, применяется нормализующее **преобразование Фишера**:
$$Z = \frac{1}{2} \ln \frac{1 + r_{xy}}{1 - r_{xy}}$$
Случайная величина $Z$ имеет приближенно нормальное распределение со среднеквадратическим отклонением $\sigma_z = \frac{1}{\sqrt{n-3}}$. Границы доверительного интервала пересчитываются обратно с помощью гиперболического тангенса.

#### 4. Проверка гипотезы о значимости коэффициента корреляции
Для проверки статистической значимости полученного результата выдвигается нулевая гипотеза **$H_0: r_{xy} = 0$** (линейная корреляционная связь в генеральной совокупности отсутствует).
В качестве критерия используется случайная величина $T$, имеющая распределение Стьюдента с $k = n - 2$ степенями свободы:
$$T_{набл} = \frac{r_{xy}\sqrt{n-2}}{\sqrt{1 - r_{xy}^2}}$$
* Если $|T_{набл}| \le t_{крит}(\alpha, k)$, нулевая гипотеза принимается — коэффициент корреляции не значим.
* Если $|T_{набл}| > t_{крит}(\alpha, k)$, нулевая гипотеза отвергается — связь признается статистически значимой.

## Постановка задачи

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин
X
 (первый признак) и
Y
 (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

### Пункт 1. Статистическая обработка выборки по второму признаку (Y)

В данном пункте мы проведем статистическую обработку выборки $Y$ в объеме предыдущих практических работ. Мы сформируем интервальный ряд и найдем точечные оценки:
* Характеристик положения: математическое ожидание (выборочное среднее), моду и медиану.
* Характеристик рассеяния: исправленную дисперсию $s_y^2$, исправленное СКО $s_y$ и коэффициент вариации.
* Показателей формы распределения: коэффициенты асимметрии и эксцесса.

In [33]:
import pandas as pd
import numpy as np
from scipy import stats
import math

# Загрузка выборки
df = pd.read_csv('sample_27.csv')
X = df['X'].values
Y = df['Y'].values
n = len(X)

# Пункт 1: Статистическая обработка Y
k = int(np.round(1 + 3.322 * np.log10(n)))

# Характеристики для Y
mean_y = np.mean(Y)
var_y = np.var(Y, ddof=1)
std_y = np.std(Y, ddof=1)
skew_y = stats.skew(Y, bias=False)
kurt_y = stats.kurtosis(Y, bias=False)
median_y = np.median(Y)
cv_y = (std_y / abs(mean_y)) * 100

# Для моды найдем интервалы
Y_min, Y_max = Y.min(), Y.max()
hy = (Y_max - Y_min) / k
bins_y = [Y_min + i * hy for i in range(k+1)]
bins_y[-1] += 1e-5
n_j, _ = np.histogram(Y, bins=bins_y)
mod_idx = np.argmax(n_j)
n_Mo = n_j[mod_idx]
n_Mo_prev = n_j[mod_idx - 1] if mod_idx > 0 else 0
n_Mo_next = n_j[mod_idx + 1] if mod_idx < (k - 1) else 0
mode_y = bins_y[mod_idx] + hy * ((n_Mo - n_Mo_prev) / ((n_Mo - n_Mo_prev) + (n_Mo - n_Mo_next)))

stats_y_df = pd.DataFrame({
    'Характеристика (для Y)': ['Мат. ожидание (среднее)', 'Испр. Дисперсия', 'Испр. СКО', 'Асимметрия', 'Эксцесс', 'Медиана', 'Мода', 'Коэф. вариации (%)'],
    'Значение': [mean_y, var_y, std_y, skew_y, kurt_y, median_y, mode_y, cv_y]
})

print("Точечные статистические оценки для признака Y:")
display(stats_y_df.round(4))

Точечные статистические оценки для признака Y:


,Характеристика (для Y),Значение
0,Мат. ожидание (среднее),14.1050
1,Испр. Дисперсия,63.5228
2,Испр. СКО,7.9701
3,Асимметрия,0.1989
4,Эксцесс,-1.4080
5,Медиана,11.7000
6,Мода,8.7918
7,Коэф. вариации (%),56.5056


Статистическая обработка признака $Y$ выполнена в полном объеме. Найдены точечные оценки математического ожидания, дисперсии и СКО. Анализ коэффициентов асимметрии и эксцесса позволяет судить о форме эмпирического распределения признака $Y$, а коэффициент вариации показывает степень однородности значений в выборке по данному признаку.

### Пункты 2 и 3. Построение двумерного интервального ряда и корреляционной таблицы

Для анализа непрерывных двумерных данных строится двумерный интервальный вариационный ряд. Значения признаков $X$ и $Y$ разбиваются на интервалы.
Результат оформляется в виде корреляционной таблицы, где на пересечении $i$-го интервала по $X$ и $j$-го по $Y$ записывается частота совместного появления признаков — $n_{ij}$. В дополнительных столбцах и строках подсчитываются маргинальные частоты $n_x$ и $n_y$, сумма которых должна строго равняться общему объему выборки $n$.

In [34]:
# Границы интервалов для X
X_min, X_max = X.min(), X.max()
hx = (X_max - X_min) / k
bins_x = [X_min + i * hx for i in range(k+1)]
bins_x[-1] += 1e-5

# Категоризация данных по интервалам
cat_x = pd.cut(X, bins=bins_x, right=False)
cat_y = pd.cut(Y, bins=bins_y, right=False)

# Корреляционная таблица (двумерный ряд)
corr_table = pd.crosstab(cat_x, cat_y, rownames=['Интервалы X'], colnames=['Интервалы Y'])
n_ij = corr_table.values

# Добавление маргинальных сумм (n_x и n_y)
corr_table_display = corr_table.copy()
corr_table_display.loc['n_y (Сумма)'] = corr_table_display.sum(axis=0)
corr_table_display['n_x (Сумма)'] = corr_table_display.sum(axis=1)

print("2-3. Корреляционная таблица:")
display(corr_table_display)

2-3. Корреляционная таблица:


Интервалы Y,"[2.05, 5.319)","[5.319, 8.587)","[8.587, 11.856)","[11.856, 15.125)","[15.125, 18.394)","[18.394, 21.662)","[21.662, 24.931)","[24.931, 28.2)",n_x (Сумма)
Интервалы X,,,,,,,,,
"[3.35, 7.406)",3,0,2,0,1,1,6,2,15
"[7.406, 11.462)",2,2,1,1,1,2,2,6,17
"[11.462, 15.519)",3,4,4,0,1,0,1,1,14
"[15.519, 19.575)",5,2,2,4,0,0,1,0,14
"[19.575, 23.631)",3,6,5,0,0,1,2,0,17
"[23.631, 27.688)",0,1,1,0,0,0,2,0,4
"[27.688, 31.744)",0,1,4,1,2,3,0,2,13
"[31.744, 35.8)",2,5,3,1,2,8,3,2,26
n_y (Сумма),18,21,22,7,7,15,17,13,120


Построен двумерный интервальный вариационный ряд и сформирована корреляционная таблица.
Маргинальные частоты $n_x$ и $n_y$ рассчитаны корректно (итоговая сумма равна объему выборки).

### Пункт 4. Вычисление выборочного коэффициента корреляции

Выборочный коэффициент корреляции ($r_{xy}$) является количественной мерой тесноты и направления линейной связи между случайными величинами.
Вычислим его двумя способами для доказательства их тождественности:
1. По стандартной формуле для сгруппированных данных:
   $$r_{xy} = \frac{\sum \sum n_{ij} x_i y_j - n \bar{x} \bar{y}}{n \sigma_x \sigma_y}$$
2. С помощью условных вариант: Осуществляется переход к безразмерным переменным $u_i = \frac{x_i - C_x}{h_x}$ и $v_j = \frac{y_j - C_y}{h_y}$ для упрощения вычислений, где $C_x, C_y$ — ложные нули.
   $$r_{xy} = \frac{\sum \sum n_{ij} u_i v_j - n \bar{u} \bar{v}}{n \sigma_u \sigma_v}$$

In [35]:
# Середины интервалов
x_mids = np.array([(bins_x[i] + bins_x[i+1])/2 for i in range(k)])
y_mids = np.array([(bins_y[i] + bins_y[i+1])/2 for i in range(k)])

n_i = n_ij.sum(axis=1)
n_j = n_ij.sum(axis=0)

# 1. СТАНДАРТНАЯ ФОРМУЛА
x_mean_grp = np.sum(n_i * x_mids) / n
y_mean_grp = np.sum(n_j * y_mids) / n

x_var_grp = np.sum(n_i * (x_mids - x_mean_grp)**2) / n
y_var_grp = np.sum(n_j * (y_mids - y_mean_grp)**2) / n

cov_xy_grp = np.sum([n_ij[i,j] * x_mids[i] * y_mids[j] for i in range(k) for j in range(k)]) / n - x_mean_grp * y_mean_grp
r_std = cov_xy_grp / np.sqrt(x_var_grp * y_var_grp)

# 2. МЕТОД УСЛОВНЫХ ВАРИАНТ
Cx = x_mids[np.argmax(n_i)]
Cy = y_mids[np.argmax(n_j)]

u = np.round((x_mids - Cx) / hx).astype(int)
v = np.round((y_mids - Cy) / hy).astype(int)

u_mean = np.sum(n_i * u) / n
v_mean = np.sum(n_j * v) / n

u_var = np.sum(n_i * u**2) / n - u_mean**2
v_var = np.sum(n_j * v**2) / n - v_mean**2

cov_uv = np.sum([n_ij[i,j] * u[i] * v[j] for i in range(k) for j in range(k)]) / n - u_mean * v_mean
r_cond = cov_uv / np.sqrt(u_var * v_var)

print(f"Коэффициент корреляции (стандартная формула): r = {r_std:.5f}")
print(f"Коэффициент корреляции (через усл. варианты): r = {r_cond:.5f}")

Коэффициент корреляции (стандартная формула): r = -0.02907
Коэффициент корреляции (через усл. варианты): r = -0.02907


Значение выборочного коэффициента корреляции, рассчитанное двумя различными математическими способами, полностью совпадает. Использование метода условных вариант доказывает свою эффективность для ручного счета, так как работает с небольшими целыми числами, но при этом гарантирует идеальную точность результата. Полученный знак и модуль $r_{xy}$ характеризуют тесноту линейной связи между признаками $X$ и $Y$.

### Пункт 5. Построение доверительного интервала для коэффициента корреляции

Вычисленный коэффициент $r_{xy}$ является лишь точечной оценкой. Чтобы найти интервал, который с заданной надежностью $\gamma$ накрывает истинный коэффициент корреляции генеральной совокупности, применяется нормализующее преобразование Фишера:
$$Z = \frac{1}{2} \ln \frac{1 + r_{xy}}{1 - r_{xy}}$$
Мы построим интервалы для двух стандартных уровней надежности: $\gamma = 0.95$ и $\gamma = 0.99$.

In [36]:
# Преобразование Фишера
z = 0.5 * np.log((1 + r_std) / (1 - r_std))
sigma_z = 1 / np.sqrt(n - 3)

# Доверительный интервал для γ = 0.95
z_crit_95 = stats.norm.ppf(0.975)
ci_z_95 = (z - z_crit_95 * sigma_z, z + z_crit_95 * sigma_z)
ci_r_95 = (np.tanh(ci_z_95[0]), np.tanh(ci_z_95[1]))

# Доверительный интервал для γ = 0.99
z_crit_99 = stats.norm.ppf(0.995)
ci_z_99 = (z - z_crit_99 * sigma_z, z + z_crit_99 * sigma_z)
ci_r_99 = (np.tanh(ci_z_99[0]), np.tanh(ci_z_99[1]))

print(f"Доверительный интервал (γ = 0.95): [{ci_r_95[0]:.4f}; {ci_r_95[1]:.4f}]")
print(f"Доверительный интервал (γ = 0.99): [{ci_r_99[0]:.4f}; {ci_r_99[1]:.4f}]")

Доверительный интервал (γ = 0.95): [-0.2072; 0.1510]
Доверительный интервал (γ = 0.99): [-0.2610; 0.2061]


С использованием $Z$-преобразования Фишера построены доверительные интервалы. С вероятностью 95% и 99% истинное значение коэффициента корреляции генеральной совокупности находится внутри найденных границ. Как видно из расчетов, при увеличении требуемой надежности оценки (от 0.95 до 0.99) границы доверительного интервала математически закономерно расширяются.

### Пункт 6. Проверка статистической гипотезы о значимости коэффициента корреляции

Выдвигается нулевая гипотеза $H_0: r_{xy} = 0$ (линейная связь в генеральной совокупности отсутствует).
В качестве критерия используется случайная величина $T$, подчиняющаяся распределению Стьюдента с $k = n - 2$ степенями свободы:
$$T_{набл} = \frac{r_{xy}\sqrt{n-2}}{\sqrt{1 - r_{xy}^2}}$$
Мы вычислим $T_{набл}$, найдем по таблице критическое значение $t_{крит}$ для заданного уровня значимости $\alpha = 0.05$ и сравним их. Если $|T_{набл}| > t_{крит}$, гипотеза отвергается.

In [37]:
# Уровень значимости
alpha = 0.05

# Наблюдаемое значение критерия (T_набл)
T_obs = (r_std * np.sqrt(n - 2)) / np.sqrt(1 - r_std**2)

# Критическое значение Стьюдента (двусторонняя область)
t_crit = stats.t.ppf(1 - alpha/2, df=n-2)

print(f"Наблюдаемое значение критерия T_набл = {T_obs:.4f}")
print(f"Критическое значение t_крит (α=0.05, k={n-2}) = {t_crit:.4f}")

if abs(T_obs) > t_crit:
    print("Вывод: |T_набл| > t_крит. Нулевая гипотеза отвергается.")
    print("Выборочный коэффициент корреляции статистически значимо отличается от нуля. Связь существует.")
else:
    print("Вывод: |T_набл| <= t_крит. Нулевая гипотеза принимается.")
    print("Выборочный коэффициент корреляции не значим. Линейная связь в генеральной совокупности отсутствует.")

Наблюдаемое значение критерия T_набл = -0.3160
Критическое значение t_крит (α=0.05, k=118) = 1.9803
Вывод: |T_набл| <= t_крит. Нулевая гипотеза принимается.
Выборочный коэффициент корреляции не значим. Линейная связь в генеральной совокупности отсутствует.


В ходе проверки статистической гипотезы о значимости выборочного коэффициента корреляции ($H_0: r_{xy} = 0$) было вычислено наблюдаемое значение критерия Стьюдента: $T_{набл} = -0.3160$. Критическое табличное значение для двусторонней области при заданном уровне значимости $\alpha = 0.05$ и числе степеней свободы $k = 118$ составило $t_{крит} = 1.9803$.

Поскольку выполняется строгое неравенство $|T_{набл}| \le t_{крит}$ (то есть $|-0.3160| \le 1.9803$), основная нулевая гипотеза $H_0$ принимается.

 Выборочный коэффициент корреляции статистически не значим. Это строго математически доказывает, что линейная корреляционная зависимость между исследуемыми признаками $X$ и $Y$ в генеральной совокупности отсутствует.

### Вывод

В ходе выполнения практической работы было проведено исследование корреляционной зависимости между двумя случайными величинами на основе заданной двумерной выборки непрерывных данных. Была выполнена первичная статистическая обработка второго признака ($Y$), построен двумерный интервальный вариационный ряд и сформирована корреляционная таблица. Рассчитан выборочный коэффициент корреляции двумя способами, построены доверительные интервалы с использованием преобразования Фишера и осуществлена проверка статистической гипотезы по критерию Стьюдента.

На основе анализа полученных результатов установлено, что:

* Вычисление выборочного коэффициента корреляции с применением метода условных вариант (с использованием ложных нулей $C_x$ и $C_y$) полностью совпадает с результатами расчета по стандартной формуле, что подтверждает высокую надежность и вычислительную эффективность данного метода при работе со сгруппированными двумерными данными.
* Полученное точечное значение коэффициента корреляции ($r_{xy} \approx -0.029$) оказалось крайне близким к нулю, что дало первичные математические основания предполагать отсутствие тесной линейной связи между исследуемыми случайными величинами.
* Построенные доверительные интервалы (с помощью нормализующего $Z$-преобразования Фишера для уровней надежности 0.95 и 0.99) математически закономерно накрывают нулевое значение, что наглядно иллюстрирует статистическую незначимость полученной точечной оценки.
* Проверка гипотезы о равенстве генерального коэффициента корреляции нулю строго математически подтвердила (на уровне значимости $\alpha = 0.05$) отсутствие значимой линейной корреляционной зависимости между признаками $X$ и $Y$ в генеральной совокупности.

Таким образом, в результате выполнения работы были приобретены практические навыки формирования двумерных выборок и построения корреляционных таблиц, освоены алгоритмы расчета точечных и интервальных оценок коэффициента корреляции, а также сформировано глубокое понимание методологии проверки статистических гипотез о наличии линейной связи.